### DISCUSSION TITLE
2.0: Is there a way to tell sqlalchemy to completely ignore an annotation?

### DISCUSSION THREAD

**Discussion by cqfd** [sqlalchemy/sqlalchemy#10510] [Usage Questions] [answered]
I'm porting a large 1.4 codebase to 2.0. We use quite a lot of our own type annotations. Despite using `__allow_unmapped__ = True`, I've run into several situations where a previously innocent, mypy-only annotation will now crash sqlalchemy's annotation inspection code.

As one example, some existing code of the form `some_col = deferred(Column(LargeBinary))` will crash sqlalchemy if you change it to `some_col: Any = deferred(Column(LargeBinary))`, with `sqlalchemy.exc.ArgumentError: Column must be constructed with a non-blank name or assign a non-blank .name before adding to a Table.`. (I've also tried switching to `mapped_column(..., deferred=True)`, but that seems to behave slightly differently/causes a bunch of tests to fail—ideally I'd leave this code as-is.)

I suppose I'm wondering about two things here:
a) Am I misconfiguring `__allow_unmapped__`? I've done a bunch of stepping through the sqlalchemy annotation introspection code in a debugger, and the flag seems to be set.
b) If not, is there a way/trick to get sqlalchemy to completely ignore a type annotation, so that it's guaranteed to only affect mypy?

Currently I'm getting around this by falling back to `.pyi` files or type comments.

**Comment by zzzeek** [comment, marked as answer]
> 
> As one example, some existing code of the form `some_col = deferred(Column(LargeBinary))` will crash sqlalchemy if you change it to `some_col: Any = deferred(Column(LargeBinary))`, with `sqlalchemy.exc.ArgumentError: Column must be constructed with a non-blank name or assign a non-blank .name before adding to a Table.`. 

well that's a bug, you should report it by itself and it will be fixed in the next release.

> 
> I suppose I'm wondering about two things here: a) Am I misconfiguring `__allow_unmapped__`? I've done a bunch of stepping through the sqlalchemy annotation introspection code in a debugger, and the flag seems to be set. 

the runtime introspection logic in SQLAlchemy 2.0 is huge and all new so you should make isolated test cases, confirm they act in an unexpected way, then then report them here so we can get them fixed or clarified.  

> b) If not, is there a way/trick to get sqlalchemy to completely ignore a type annotation, so that it's guaranteed to only affect mypy?

if you really have to do this you can use TYPE_CHECKING blocks, which is less effort than pyi files

**Reply by cqfd** [reply]
Ok will do—and that's a good point about `TYPE_CHECKING`, somehow I totally forgot about that option 🤦

**Reply by zzzeek** [reply]
I've already made a fix for the Any issue, were you making a github issue for it?  I can just make one if not

**Reply by cqfd** [reply]
Ah cool! Go for it, you're going to beat me to it—I'll see if I can find any other issues to file this weekend though. Thanks for the help, this upgrade to 2.0 is going to be such a big quality of life improvement to the whole team 🙏
